src/utils/encoding/cid_de_cbor.rs

```diff
@@ -0,0 +1,247 @@
+                E: de::Error,
+            {
+                Ok(())
+            }
+
+            #[inline]
+            fn visit_i128<E>(self, _v: i128) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(())
+            }
+
+            #[inline]
+            fn visit_f64<E>(self, _v: f64) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(())
+            }
+
+            #[inline]
+            fn visit_bool<E>(self, _v: bool) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(())
+            }
+
+            #[inline]
+            fn visit_none<E>(self) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(())
+            }
+        }
+
+        deserializer.deserialize_any(FilterCidsVisitor(self.0))
+    }
+}
+
+impl<'de> de::Deserialize<'de> for CidVec {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let mut vec = CidVec(Vec::new());
+        FilterCids(&mut vec.0).deserialize(deserializer)?;
+        Ok(vec)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::ipld::DfsIter;
+
+    use crate::utils::encoding::extract_cids;
+    use cid::multihash::Code::Blake2b256;
+    use cid::multihash::MultihashDigest;
+    use cid::Cid;
+
+    use fvm_ipld_encoding::DAG_CBOR;
+    use libipld_core::ipld::Ipld;
+    use quickcheck::{Arbitrary, Gen};
+    use quickcheck_macros::quickcheck;
+
+    #[derive(Debug, Clone)]
+    pub struct IpldWrapper {
+        inner: Ipld,
+    }
+
+    impl Arbitrary for IpldWrapper {
+        fn arbitrary(g: &mut Gen) -> Self {
+            let mut ipld = Ipld::arbitrary(g);
+
+            fn cleanup_ipld(ipld: &mut Ipld, g: &mut Gen) {
+                match ipld {
+                    // [`Cid`]s have to be valid in order to be decodable.
+                    Ipld::Link(cid) => {
+                        *cid = Cid::new_v1(
+                            DAG_CBOR,
+                            Blake2b256.digest(&[
+                                u8::arbitrary(g),
+                                u8::arbitrary(g),
+                                u8::arbitrary(g),
+                            ]),
+                        )
+                    }
+                    Ipld::Map(map) => map.values_mut().for_each(|val| cleanup_ipld(val, g)),
+                    Ipld::List(vec) => vec.iter_mut().for_each(|val| cleanup_ipld(val, g)),
+                    // Cleaning up Integer and Float in order to avoid parser mistakes that result
+                    // in tag detection and a subsequent Cid decoding failure.
+                    // See https://github.com/ipld/serde_ipld_dagcbor/blob/master/src/de.rs#L178 and
+                    // https://github.com/ipld/serde_ipld_dagcbor/blob/master/src/de.rs#L119 .
```

Comment:
This could use more explanation, for someone fairly fresh to this subject it's a bit shady to replace all numeric occurrences with zeroes.

Also, let's use permalinks, otherwise those links will get quickly outdated and point to random pieces of code (or 404).

Reply:
Sounds good.

Reply:
done